compiler/damlc/daml-stdlib-src/DA/Internal/Exception.daml

```diff
@@ -0,0 +1,133 @@
+    , HasMessage e
+    , HasToAnyException e
+    , HasFromAnyException e
+    )
+
+-- | Part of the `Exception` constraint.
+class HasThrow e where
+    -- | Throw exception in a pure context.
+    throwPure : forall t. e -> t
+
+-- | Part of the `Exception` constraint.
+class HasMessage e where
+    -- | Get the error message associated with an exception.
+    message : e -> Text
+
+-- | Part of the `Exception` constraint.
+class HasToAnyException e where
+    -- | Convert an exception type to AnyException.
+    toAnyException : e -> AnyException
+
+-- | Part of the `Exception` constraint.
+class HasFromAnyException e where
+    -- | Convert an AnyException back to the underlying exception type, if possible.
+    fromAnyException : AnyException -> Optional e
+
+--------------------------------------------------------------
+
+-- No throw for AnyException.
+
+instance HasMessage AnyException where
+    message = undefined -- primitive @"BEAnyExceptionMessage"
+
+instance HasToAnyException AnyException where
+    toAnyException e = e
+
+instance HasFromAnyException AnyException where
+    fromAnyException = Some
+
+--------------------------------------------------------------
+
+instance HasThrow GeneralError where
+    throwPure = undefined -- primitive @"EThrow"
+
+instance HasMessage GeneralError where
+    message = undefined -- primitive @"BEGeneralErrorMessage"
+
+instance HasToAnyException GeneralError where
+    toAnyException = undefined -- primitive @"EToAnyException"
+
+instance HasFromAnyException GeneralError where
+    fromAnyException = undefined -- primitive @"EFromAnyException"
+
+--------------------------------------------------------------
+
+instance HasThrow ArithmeticError where
+    throwPure = undefined -- primitive @"EThrow"
+
+instance HasMessage ArithmeticError where
+    message = undefined -- primitive @"BEArithmeticErrorMessage"
+
+instance HasToAnyException ArithmeticError where
+    toAnyException = undefined -- primitive @"EToAnyException"
+
+instance HasFromAnyException ArithmeticError where
+    fromAnyException = undefined -- primitive @"EFromAnyException"
+
+--------------------------------------------------------------
+
+instance HasThrow ContractError where
+    throwPure = undefined -- primitive @"EThrow"
+
+instance HasMessage ContractError where
+    message = undefined -- primitive @"BEContractErrorMessage"
+
+instance HasToAnyException ContractError where
+    toAnyException = undefined -- primitive @"EToAnyException"
+
+instance HasFromAnyException ContractError where
+    fromAnyException = undefined -- primitive @"EFromAnyException"
+
+--------------------------------------------------------------
+
+-- | Action type in which `throw` is supported.
+class Action m => ActionThrow m where
+    throw : Exception e => e -> m t
+
+-- | Action type in which `try ... catch ...` is supported.
+class ActionThrow m => ActionCatch m where
+    -- | Handle an exception. Use the `try ... catch ...` syntax
```

Comment:
Should we just hide this? Can we hide a method from a typeclass in daml-docs?

Reply:
We can. Hiding is an option, but I think it might be better to expose it and let people know that there's a better thing to use, and that they only need to think about this method if they want to implement their own instance of ActionCatch.

Reply:
sounds good, I guess I didn’t expect anyone to write their own instance but thinking about it, I think we both have at least one user in mind who will do that :slightly_smiling_face: